File: src/sys/SysPlanilha.java
package sys;

import java.util.HashMap;
import java.util.LinkedList;

public class SysPlanilha {

	private HashMap<String,Aluno> alunos = new HashMap<String,Aluno>();
	
	public SysPlanilha() {
		this.carregarDadosBasicos();
	}
	
	public void cadastrarAluno(Aluno aluno) throws AlunoJaCadastradoException{
		try {
			this.pesquisarAlunoPorMatricula(aluno.getMatricula());
		} catch (AlunoNaoCadastradoException e) {
			this.alunos.put(aluno.getMatricula(), aluno);
		}
	}
	
	public Aluno pesquisarAlunoPorMatricula(String matricula) throws AlunoNaoCadastradoException{
		
		if(!this.alunos.containsKey(matricula)) {
			this.carregarAluno(matricula);
		}
		return this.alunos.get(matricula);
	}
	
	public LinkedList<Aluno> pesquisarAlunosPorNome(String nome) throws AlunoNaoCadastradoException{
		
		LinkedList<Aluno> temp = new LinkedList<Aluno>();
		
		for(Aluno a: this.alunos.values()) {
			if(a.getNome().contains(nome)) {
				temp.add(a);
			}
		}
		for(Aluno a: this.carregarAlunosPorNome(nome)) {
			temp.add(a);
		}
		
		return temp;
	}
	
	public void editarNotaDoAluno(String matricula, NomesMateria materia, double nota, int unidade) throws AlunoNaoCadastradoException{
		this.alunos.get(matricula).cadastrarNota(materia, nota, unidade);
	}
	
	public void excluirAluno(String matricula) throws AlunoNaoCadastradoException{
		this.pesquisarAlunoPorMatricula(matricula);
		this.alunos.remove(matricula);
	}
	
	public void carregarDadosBasicos() {
		//TODO: implementar o load dos dados de matriculas
	}
	
	public void carregarAluno(String matricula)  throws AlunoNaoCadastradoException{
		//TODO: imlementar o lazyload de alunos
	}
	
	public LinkedList<Aluno> carregarAlunosPorNome(String nome) {
		return null;//TODO: imlementar o load de alunos
	}
	
	public void salvarDados() {
		//TODO: implementar o save de todos os dados em cache
	}

}

File: srcPython/SYS/SysPlanilha.py
from SYS.Aluno import Aluno
from SYS.Materias import NomesMateria
import pickle
from os.path import exists as pathExists

class AlunoJaCadastradoException(Exception):
	def __init__(self,message):
		self.message = message

class AlunoNaoCadastradoException(Exception):
	def __init__(self,message):
		self.message = message

class SysPlanilha:
	
	ARQUIVO_DE_DADOS  = "arquivos.pkl"

	def __init__(self):
		self.alunos = dict() #{"matricula":Aluno}
		if pathExists(self.ARQUIVO_DE_DADOS):
			self.carregarDados()

	def cadastrarAluno(self, aluno:Aluno):
		try:
			self.pesquisarAlunoPorMatricula(aluno.matricula)
			raise AlunoJaCadastradoException(f"Já existe um aluno com a matrícula {aluno.matricula} cadastrada.")
		except AlunoNaoCadastradoException:
			self.alunos[aluno.matricula] = aluno

	def pesquisarAlunoPorMatricula(self, matricula:str):
		if matricula not in self.alunos.keys():
			raise AlunoNaoCadastradoException(f"Não existe nenhum aluno no banco de dados com a matrícula {matricula}")
		else:
			return self.alunos[matricula]

	def editarNotaDoAluno(self, matricula:str, materia:NomesMateria, unidade:int, nota:float):
		self.pesquisarAlunoPorMatricula(matricula)
		self.alunos[matricula].cadastrarNota(materia, unidade, nota)

	def excluirAluno(self, matricula:str):
		self.pesquisarAlunoPorMatricula(matricula)
		del self.alunos[matricula]

	def salvarDados(self):
		#criando o dicinário#
		dados = {#aqui estou criando 2 palavaras chaves para guardar as variaveis que irão ser salvas nos arquivos
		 	"quantMatriculas":Aluno.getMatriculaCount(), #variavel estatica
		 	"alunosList":self.alunos,
		 	}
		 	#abrindo um aquivo para salvar os dados# o With ele fecha automaticamente a stream de dados após a execução do bloco
		with open(self.ARQUIVO_DE_DADOS,"wb") as arquivo: # wb = whrite bytes 
			pickle.dump(dados, arquivo) #escreve dentro da stream, na variável arquivo, os dados
	#		
	def carregarDados(self):
		with open(self.ARQUIVO_DE_DADOS,"rb") as arquivo: #abri o arquivo e ler o arquivo com rb = readbyte
			dados = pickle.load(arquivo) #aqui esta sendo carregado os arquivos para variavel dados
			self.alunos = dados["alunosList"] #ele estar pegado do dicionario todos os dados dos alunos
			quantMatriculas = dados["quantMatriculas"] #aqui estar sendo pega a quantidade de matriculas
		Aluno.setMatriculaCount(quantMatriculas)#está passando para variável estática da classe aluno a quantidade de matrícula
		